### PR TITLE
Simpleem

### PR DESCRIPTION
**Pull request by redhog**
#### Summary
Merge in https://github.com/emerald-geomodelling/experimental-simpeg-ext, see this repo for more details and usage, as well as the notebooks in the docs subdirectory of this PR.

#### PR Checklist
* [ ] If this is a work in progress PR, set as a Draft PR
* [ ] Linted my code according to the [style guides](https://docs.simpeg.xyz/content/basic/practices.html#style).
* [ ] Added [tests](https://docs.simpeg.xyz/content/basic/practices.html#testing) to verify changes to the code.
* [ ] Added necessary documentation to any new functions/classes following the
      expect [style](https://docs.simpeg.xyz/content/basic/practices.html#documentation).
* [ ] Added relevant method tags (i.e. `GRAV`, `EM`, etc.)
* [ ] Marked as ready for review (ff this is was a draft PR), and converted 
      to a Pull Request
* [ ] Tagged ``@simpeg/simpeg-developers`` when ready for review.

### CODE REVIEW

**Comment by sgkang** [comment]
thanks @redhog; I'll work on this next week.

**Comment by jcapriot** [comment]
Very interesting thanks for you contributions! There's a lot of stuff here that could be expanded to be useful to more than just the EM1D codes.

I'm not sure of the merge target though, I'm don't know whether or not the `em1d_stitched` branch will be merged itself and it might end up being superseded by more recent items in the `main` branch. I would recommend rebasing the changes on the main branch to compare there.

**Comment by redhog** [comment]
@jcapriot The target was chosen specifically because we internally use this branch for SCI AEM inversion, which is what my code initially targets. I'd be happy to rebase this, at least if `main` will later be merged into `em1d_stitched`, but I'd defer this to @sgkang who knows more what's going on in the different branches...

I can easily see this expanded to other EM scenarios than what we do (AEM), like TTEM etc, but you think this has application outside EM even? I'm not familiar enough with other methods, and I kinda assumed that for seismic etc the instrument setup was arbitrary for each sounding (random placement of geophones)?

**Comment by redhog** [comment]
@jcapriot So the main idea is this:

Given the assumption that the instrument setup doesn't change between soundings, a lot of simplifications can be made. In particular, we can describe the sounding data as an array of shape (NSoundings, MMeasurements), together with an array of coordinates (x, y, ztopography, zaltitude). For EM data, we read these arrays using the libaarhusxyz library from files in the XYZ format used by SkyTEM.

The system itself, together with inversion parameters can then be described in a python class, where each SimPEG object, such as Survey, InvProb, Receiver etc, is created by a method. This class is instantiated with the sounding data above as a parameter, to form an object on which inversion can be performed. But the same class can also be instantiated with a model instead of data, and a subset of the same methods used to construct what is needed to do forward modelling.

To what degree does this separation apply to other problems?

**Comment by jcapriot** [comment]
The pieces of the code the create and return the `Survey` object for the loaded in data is fairly independent of the 1D code. The sources and receivers used in the 1D code are the same sources and receivers used in the 3D code. So, with very little modification (I think you essentially do not need to set an `i_sounding` property), the same survey could be used for multiple different TDEM simulations.

**Comment by redhog** [comment]
Cool! I'd be happy to help generalize the base class of this to make this possible :)

Do you have some example notebook that performs 3D inversion that I could look at to see how it maps?

File: SimPEG/electromagnetics/utils/static_instrument/base.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+            inverse_problem.BaseInvProblem(
+                self.make_misfit(thicknesses),
+                self.make_regularization(thicknesses),
+                self.make_optimizer()),
+            self.make_directives())
+
+    def make_forward(self):
+        return self.make_simulation(self.make_survey(), self.make_thicknesses())
+        
+    def inverted_model_to_xyz(self, model, thicknesses):
```
Comment (sgkang) [inline SimPEG/electromagnetics/utils/static_instrument/base.py:243]: Nice!

File: SimPEG/optimization.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 import scipy.sparse as sp
 from six import string_types
 
-from .utils.solver_utils import SolverWrapI, Solver
+from .utils.solver_utils import SolverWrapI, Solver, SolverDiag
```
[the remark this answers is not in the document]
Comment (sgkang) [reply SimPEG/optimization.py:7]: image

File: SimPEG/electromagnetics/utils/static_instrument/utils.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+from SimPEG.utils import mkvc
+import SimPEG.electromagnetics.time_domain as tdem
+import SimPEG.electromagnetics.utils.em1d_utils
+from SimPEG.electromagnetics.utils.em1d_utils import get_2d_mesh,plot_layer, get_vertical_discretization_time
+from SimPEG.regularization import LaterallyConstrained, RegularizationMesh
+
+import scipy.stats
+
+
+def make_2layer(xdist, dtb, layers, res_upper=30, res_lower=300, x=None, y=None):
```
Comment (sgkang) [inline SimPEG/electromagnetics/utils/static_instrument/utils.py:36]: add docstrings

File: SimPEG/electromagnetics/utils/static_instrument/utils.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    xyz = libaarhusxyz.XYZ()
+
+    xyz.flightlines = flightlines
+    xyz.layer_data["resistivity"] = resistivity
+    xyz.layer_data["dep_top"] = pd.DataFrame(gzt.T)
+    xyz.layer_data["dep_bot"] = pd.DataFrame(gzb.T)
+    
+    return xyz
+
+def add_noise(xyz, rel_uncertainty=0.01):
```
Comment (sgkang) [inline SimPEG/electromagnetics/utils/static_instrument/utils.py:64]: add docstrings

File: SimPEG/electromagnetics/utils/static_instrument/utils.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    dpred = xyz.dbdt_ch1gt.values.flatten()
+    noise = rel_uncertainty*np.abs(dpred)*np.random.rand(len(dpred))
+    xyz.layer_data["dbdt_ch1gt"] += noise.reshape(xyz.dbdt_ch1gt.shape)
+
+def add_uncertainty_normal(xyz, rel_uncertainty):
```
Comment (sgkang) [inline SimPEG/electromagnetics/utils/static_instrument/utils.py:69]: Not sure what is this for?

File: SimPEG/electromagnetics/utils/static_instrument/base.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+            # reg.get_grad_horizontal(self.xyz.flightlines[["x", "y"]], hz, dim=2, use_cell_weights=True)
+            # ps, px, py = 0, 0, 0
+            # reg.norms = np.c_[ps, px, py, 0]
+            reg.mref = np.log(np.ones(self.n_param(thicknesses)) * 1/self.start_res)
+            # reg.mrefInSmooth = False
+            return reg
+        else:
+            coords = self.xyz.flightlines[[self.xyz.x_column, self.xyz.y_column]].astype(float).values
+            # FIXME: Triangulation fails if all coords are on a line, as in a typical synthetic case...
+            # Ideally, shouldn't this degrade gracefully to an LCI?
```
Comment (sgkang) [inline SimPEG/electromagnetics/utils/static_instrument/base.py:191]: Yes, it does. You can still invert all lines and just put constraints along the line by adjusting edge parameters. I'll provide an example code later.

File: SimPEG/electromagnetics/utils/static_instrument/dual.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+from SimPEG.utils import mkvc
+import SimPEG.electromagnetics.time_domain as tdem
+import SimPEG.electromagnetics.utils.em1d_utils
+from SimPEG.electromagnetics.utils.em1d_utils import get_2d_mesh,plot_layer, get_vertical_discretization_time
+from SimPEG.regularization import LaterallyConstrained, RegularizationMesh
+
+import scipy.stats
+from . import base
+
+class DualMomentTEMXYZSystem(base.XYZSystem):
```
Comment (sgkang) [inline SimPEG/electromagnetics/utils/static_instrument/dual.py:36]: Lovely!

**Review by sgkang — COMMENTED** [review]
Hi Egil, Thank you so much for your contribution! This looks great overall. 
I ran notebooks examples in: https://github.com/emerald-geomodelling/experimental-simpeg-ext and I was able to run them. 

There are lots of things, and it would be beneficial to have a chat to brainstorm how to bring stuff in to simpeg.

File: SimPEG/electromagnetics/utils/static_instrument/base.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+                alpha_z = 1.)
+            # reg.get_grad_horizontal(self.xyz.flightlines[["x", "y"]], hz, dim=2, use_cell_weights=True)
+            # ps, px, py = 0, 0, 0
+            # reg.norms = np.c_[ps, px, py, 0]
+            reg.mref = np.log(np.ones(self.n_param(thicknesses)) * 1/self.start_res)
+            # reg.mrefInSmooth = False
+            return reg
+        else:
+            coords = self.xyz.flightlines[[self.xyz.x_column, self.xyz.y_column]].astype(float).values
+            # FIXME: Triangulation fails if all coords are on a line, as in a typical synthetic case...
```
[the remark this answers is not in the document]
Comment (redhog) [reply SimPEG/electromagnetics/utils/static_instrument/base.py:190]: Haha, yes :) I guess this should make it into a separate ticket to solve?

File: SimPEG/electromagnetics/utils/static_instrument/base.py [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+            reg = SimPEG.regularization.LaterallyConstrained(
+                mesh_reg, mapping=reg_map,
+                alpha_s = self.alpha_s,
+                alpha_r = self.alpha_r,
+                alpha_z = self.alpha_z,
+            )
+            reg.mref = np.log(np.ones(self.n_param(thicknesses)) * 1/self.start_res)
+            return reg
+            
+    def make_directives(self):
```
[the remark this answers is not in the document]
Comment (redhog) [reply SimPEG/electromagnetics/utils/static_instrument/base.py:210]: You mean do something like

```
regargs = {}
for name in dir(self):
    if name.startswith("reg_"):
        regargs[name[4:]] = getattr(self, name)
reg = SimPEG.regularization.LaterallyConstrained(
                mesh_reg, mapping=reg_map, **regargs)
```

Not a bad idea :) The exact parameters I expose on the system object is a bit arbitrary for now; it's basically what we happened to need for our experiments...

File: SimPEG/electromagnetics/utils/static_instrument/utils.py (hunk 5 above)
Comment (redhog) [reply SimPEG/electromagnetics/utils/static_instrument/utils.py:69]: The add uncertainty  functions are for when you're generating synthetic data, and want to add a "measured" uncertainty (as if based on stddev of stacked actual measurements). This one adds random uncertainty scaled by the data.  add_uncertainty() below, adds a fixed uncertainty scaled by the data.

File: SimPEG/optimization.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (redhog) [reply SimPEG/optimization.py:7]: haha, that's a bugfix. In the version of SimPEG I had checked out when I wrote this, this import was missing, but used in optimization.py, so it failed to run inversions...

**Comment by redhog** [comment]
Question: *.ipynb is currently in .gitignore for SimPEG. I have two jupyter notebooks that showcase how to use this code. Should I commit them in a docs subfolder anyway, or what's the preferred way to do tutorials?

**Comment by jcapriot** [comment]
We don't have notebook files in the repository. However, you can download the tutorials on the documentation page as jupyter notebooks. These files are written as .py scripts in their respective `tutorials/...` directory and then get rendered by the `sphinx_gallery` extension into both rst pages for the documnetation and jupyter notebook files for distribution.

So essentially, convert your jupyter notebook to a python script, add a little bit of formatting, and put it in the tutorials folder.

**Comment by redhog** [comment]
@jcapriot How do you render a tutorial w/o having to build/rerender all the docs? Building all of them seems to take forever for me....

**Comment by jcapriot** [comment]
I'm going to close this in favor of #1220, since that is stated to be an updated version of this PR. If you still need this open feel free to re-open it.

**Comment by redhog** [comment]
No, please close. But I think we should try to merge https://github.com/simpeg/simpeg/pull/1220 and also try to cherrypick or merge from there to master...

**Pull request closed without merging**